url85940d3d8e_ccp999_Expr4j_tgz-pJ8-org_boris_expr_util_ExprEvaluatorJ8/src/org/boris/expr/engine/IterationEngine.java

```diff
@@ -0,0 +1,138 @@
+/*******************************************************************************
+ * This program and the accompanying materials
+ * are made available under the terms of the Common Public License v1.0
+ * which accompanies this distribution, and is available at 
+ * http://www.eclipse.org/legal/cpl-v10.html
+ * 
+ * Contributors:
+ *     Peter Smith
+ *******************************************************************************/
+package org.boris.expr.engine;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+
+import org.boris.expr.Expr;
+import org.boris.expr.ExprArray;
+import org.boris.expr.ExprEvaluatable;
+import org.boris.expr.ExprException;
+import org.boris.expr.ExprNumber;
+
+public class IterationEngine extends AbstractCalculationEngine
+{
+    private int maxIterations = 100;
+    private double maxChange = 0.0001;
+    private Map<Range, Expr> inputExprs = new HashMap();
+
+    public IterationEngine(EngineProvider provider) {
+        super(provider);
+    }
+
+    public void setMaxIterations(int iterations) {
+        this.maxIterations = iterations;
+    }
+
+    public void setMaxChange(double change) {
+        this.maxChange = Math.abs(change);
+    }
+
+    public void calculate(boolean force) throws ExprException {
+        if (autoCalculate && !force)
+            return;
+
+        calc();
+    }
+
+    private void calc() throws ExprException {
+        Set<Range> inputs = getInputRanges();
+        Map<Range, Expr> valueChanges = new HashMap();
+        for (int i = 0; i < maxIterations; i++) {
+            double change = 0;
+            for (Range r : inputs) {
+                Expr e = inputExprs.get(r);
+                if (e instanceof ExprEvaluatable) {
+                    e = ((ExprEvaluatable) e).evaluate(this);
+                    values.put(r, e);
+                    valueChanges.put(r, e);
+                    double c = findChange(values.get(r), e);
+                    if (c > change)
+                        change = c;
+                }
+            }
+
+            if (change < maxChange)
+                break;
+        }
+
+        for (Range r : valueChanges.keySet()) {
+            provider.valueChanged(r, valueChanges.get(r));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: Iterating over [map.keySet()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#keySet--) and calling [map.get(Key)](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#get-java.lang.Object-) to get the values within the loop is inefficient. 
**Fix**: Iterate over [map.entrySet()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#entrySet--) and access the values using [entry.getValue()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.Entry.html#getValue--) to avoid potential performance penalty due to lookup.